modules/core/src/main/java/org/apache/ignite/internal/jdbc2/lob/JdbcBlobMemoryStorage.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.jdbc2.lob;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+/**
+ * Heap memory based read-write implementation of {@link JdbcBlobStorage}.
+ *
+ * <p>Keeps data in list of byte array buffers (of different size) to avoid memory
+ * reallocation and data coping on write operations (which append data in particular).
+ */
+class JdbcBlobMemoryStorage implements JdbcBlobStorage {
+    /** Max capacity when it is still reasonable to double size for new buffer. */
+    private static final int MAX_CAP = 32 * 1024 * 1024;
+
+    /** The list of buffers. */
+    private List<byte[]> buffers = new ArrayList<>();
+
+    /** The total number of bytes in all buffers. */
+    private long totalCnt;
```

Comment:
We know for sure that `totalCnt <= MAX_ARRAY_SIZE`, isn't it?
So, we want to use `int` variables to track position and data length everywhere internally.

Let's use `int` here and in other places, it should simplify code.